shinysdr/plugins/limesdr.py

```diff
@@ -0,0 +1,317 @@
+            gr.io_signature(1, 1, gr.sizeof_gr_complex * 1),
+        )
+        
+        self.__source = source
+        self.__name = name
+        self.__tuning = tuning
+        
+        self.connect(self.__source, self)
+        
+        # State of the source that there are no getters for, so we must keep our own copy of
+        self.__track_gain = 50.
+        source.set_gain(int(self.__track_gain), ch)
+
+        self.__track_bandwidth = max(sample_rate / 2, 1.5e6)
+        source.set_analog_filter(True, self.__track_bandwidth, ch)
+
+        self.__lna_path_type = EnumT({
+            LNANONE: 'None',
+            LNAH: 'LNAH',
+            LNAL: 'LNAL',
+            LNAW: 'LNAW',
+        })
+        if device_type == LimeSDRMini:
+            self.__lna_path_type = EnumT({
+                LNAH: 'LNAH',
+                LNAW: 'LNAW',
+            })
+        self.__track_lna_path = lna_path
+        
+        # Blocks
+        self.__state_while_inactive = {}
+        self.__placeholder = blocks.vector_source_c([])
+        
+        self.__signal_type = SignalType(
+            kind='IQ',
+            sample_rate=sample_rate)
+        self.__usable_bandwidth = tuning.calc_usable_bandwidth(sample_rate)
+    
+    @exported_value(type=SignalType, changes='never')
+    def get_output_type(self):
+        return self.__signal_type
+    
+    # implement IRXDriver
+    def get_tune_delay(self):
+        return DEFAULT_DELAY
+
+    # implement IRXDriver
+    def get_usable_bandwidth(self):
+        if self.__track_bandwidth:
+            return self.__tuning.calc_usable_bandwidth(self.__track_bandwidth)
+        return self.__usable_bandwidth
+    
+    # implement IRXDriver
+    def close(self):
+        self._stop_rx()
+        self.__tuning = None
+
+    @exported_value(
+        type_fn=lambda self: self.__lna_path_type,
+        changes='this_setter',
+        label='LNA Path')
+    def get_lna_path(self):
+        return self.__track_lna_path
+
+    @setter
+    def set_lna_path(self, lna_path):
+        self.__track_lna_path = int(lna_path)
+        self.__source.set_lna_path(int(lna_path), ch)
+    
+    @exported_value(
+        type_fn=lambda self: RangeT([(0, 70)], unit=units.dB, strict=False),
+        changes='this_setter',
+        label='Gain')
+    def get_gain(self):
+        if self.__source is None: return 0.0
+        return self.__track_gain
+    
+    @setter
+    def set_gain(self, value):
+        self.__track_gain = int(value)
+        self.__source.set_gain(int(value), ch)
+
+    # zero means no filter
+    @exported_value(
+        type_fn=lambda self: RangeT([(0, 0), (1e3, min(130e6, self.__signal_type.get_sample_rate()))], unit=units.Hz),
+        changes='this_setter',
+        label='Hardware filter',
+        description='Bandwidth of the analog and digital filters.')
+    def get_bandwidth(self):
+        # Analog filter goes down to 1.5e6, digital filter goes arbitrarily low.
```

Comment:
This comment seems to belong more in `set_bandwidth`.

Reply:
Done.